pmlc/transforms/make_entrypoints.cc

```diff
@@ -0,0 +1,160 @@
+// Copyright 2020 Intel Corporation
+
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+
+#include "pmlc/transforms/pass_detail.h"
+#include "pmlc/util/ids.h"
+
+namespace LLVM = mlir::LLVM;
+using LLVMType = LLVM::LLVMType;
+
+namespace pmlc::transforms {
+namespace {
+
+class MakeEntrypointsPass final
+    : public MakeEntrypointsPassBase<MakeEntrypointsPass> {
+public:
+  void runOnOperation() final;
+
+private:
+  mlir::Location getLoc() { return getOperation().getLoc(); }
+  void initMembers();
+  mlir::FuncOp getNetworkEntry();
+
+  void makePlaidmlInit(mlir::OpBuilder &builder);
+  void makePlaidmlExecute(mlir::OpBuilder &builder, mlir::FuncOp entryFunc);
+  void makePlaidmlFini(mlir::OpBuilder &builder);
+
+  LLVMType llvmInt32Type;
+  LLVMType llvmPtrTy;
+  LLVMType llvmVoidTy;
+};
+
+void MakeEntrypointsPass::runOnOperation() {
+  mlir::FuncOp entryFunc = getNetworkEntry();
+  if (!entryFunc) {
+    return;
+  }
+
+  initMembers();
+  mlir::OpBuilder builder(getOperation().getBody()->getTerminator());
+  makePlaidmlInit(builder);
+  makePlaidmlExecute(builder, entryFunc);
+  makePlaidmlFini(builder);
```

Comment:
We should check to see if these functions are already defined or at least fail if they're already defined.